Apps/NeuroscanAcquireLSL/nsatransporter.cpp

```diff
@@ -0,0 +1,113 @@
+#include "nsatransporter.h"
+
+NsaTransporter::NsaTransporter() {}
+
+void NsaTransporter::init() {
+    npackets = 0;
+    socket = new QTcpSocket;
+    connect(socket, SIGNAL(readyRead()), this, SLOT(data_handler()));
+    connect(socket, SIGNAL(error(QAbstractSocket::SocketError)), this,
+            SLOT(error_handler(QAbstractSocket::SocketError)));
+    socket->setSocketOption(QAbstractSocket::LowDelayOption, 1);
+    timer = new QTimer;
+    timer->setInterval(100);
+    connect(timer, SIGNAL(timeout()), this, SLOT(ping()));
+    return;
+}
+
+void NsaTransporter::send(QByteArray data) {
+    socket->write(data);
+    return;
+}
+
+void NsaTransporter::set_host(const QString& host, const int& port) {
+    // qDebug()<<host<<port;
+    if (port == 0) {
+        socket->disconnectFromHost();
+        socket->waitForDisconnected();
+        timer->stop();
+        delete lsloutlet;
+        qInfo() << "disconnect from " << socket->peerName() << socket->peerPort();
+        return;
+    }
+    socket->connectToHost(host, port);
+    qInfo() << "connect to" << host << port;
+    return;
+}
+
+void NsaTransporter::data_handler() {
+    // qDebug()<<socket->readAll();
+    // read header
+    if (!packet.isValid() && socket->bytesAvailable() >= packet.headerSize()) {
+        packet.unserialize(socket->read(packet.headerSize()));
+    }
+    // read data
+    if (packet.isValid()) {
+        // read and process data packet
+        if (packet.dataSize() != 0 && socket->bytesAvailable() >= packet.dataSize()) {
+            packet.getData() = socket->read(packet.dataSize());
+
+            // process settings packet
+            if (packet.isInfo()) {
+                info = NsaBasicInfo(packet.getData());
+                auto chunk = vector<vector<qint32>>(
```

Comment:
Keeping the data in a contiguous int32 array and pushing it via push_chunk is much faster